logstash-core/lib/logstash/agent_plugin_registry.rb

```diff
@@ -0,0 +1,66 @@
+# encoding: utf-8
+require "logstash/agent"
+
+module LogStash module AgentPluginRegistry
+  DEFAULT_AGENT_NAME = :default
+  class DuplicatePluginError < Error; end
+
+  REGISTRY = {}
+
+  # Reset plugin registry to just the default plugin
+  def self.reset!
+    REGISTRY.clear
+    REGISTRY[DEFAULT_AGENT_NAME] = LogStash::Agent
+  end
+
+  reset!
```

Comment:
@colinsurprenant not able to find your comment now, but will try to explain my reasoning. I find this usage of singleton implementation here strange, the code is basically within a module, but keeping an state that need an initialize method to be run like actually @andrewvc did here. 

My problem goes with having a pattern of knowing an encapsulating there the change of state happen. If we want to follow proper object orientation this should only happen in methods.

I think this should be changed to be a class and not a module. Unless I'm obviously missing why this should be a module.